staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go

```diff
@@ -139,3 +145,86 @@ type CustomResourceDefinitionList struct {
+	Ref                  string                     `json:"$ref,omitempty"`
+	Description          string                     `json:"description,omitempty"`
+	Type                 StringOrArray              `json:"type,omitempty"`
+	Format               string                     `json:"format,omitempty"`
+	Title                string                     `json:"title,omitempty"`
+	Default              interface{}                `json:"default,omitempty"`
+	Maximum              *float64                   `json:"maximum,omitempty"`
+	ExclusiveMaximum     bool                       `json:"exclusiveMaximum,omitempty"`
+	Minimum              *float64                   `json:"minimum,omitempty"`
+	ExclusiveMinimum     bool                       `json:"exclusiveMinimum,omitempty"`
+	MaxLength            *int64                     `json:"maxLength,omitempty"`
+	MinLength            *int64                     `json:"minLength,omitempty"`
+	Pattern              string                     `json:"pattern,omitempty"`
+	MaxItems             *int64                     `json:"maxItems,omitempty"`
+	MinItems             *int64                     `json:"minItems,omitempty"`
+	UniqueItems          bool                       `json:"uniqueItems,omitempty"`
+	MultipleOf           *float64                   `json:"multipleOf,omitempty"`
+	Enum                 []interface{}              `json:"enum,omitempty"`
+	MaxProperties        *int64                     `json:"maxProperties,omitempty"`
+	MinProperties        *int64                     `json:"minProperties,omitempty"`
+	Required             []string                   `json:"required,omitempty"`
+	Items                *JSONSchemaPropsOrArray    `json:"items,omitempty"`
+	AllOf                []JSONSchemaProps          `json:"allOf,omitempty"`
+	OneOf                []JSONSchemaProps          `json:"oneOf,omitempty"`
+	AnyOf                []JSONSchemaProps          `json:"anyOf,omitempty"`
+	Not                  *JSONSchemaProps           `json:"not,omitempty"`
+	Properties           map[string]JSONSchemaProps `json:"properties,omitempty"`
+	AdditionalProperties *JSONSchemaPropsOrBool     `json:"additionalProperties,omitempty"`
+	PatternProperties    map[string]JSONSchemaProps `json:"patternProperties,omitempty"`
+	Dependencies         JSONSchemaDependencies     `json:"dependencies,omitempty"`
+	AdditionalItems      *JSONSchemaPropsOrBool     `json:"additionalItems,omitempty"`
+	Definitions          JSONSchemaDefinitions      `json:"definitions,omitempty"`
+}
+
+// JSONSchemaURL represents a schema url. Defaults to JSON Schema Specification Draft 4.
+type JSONSchemaURL string
+
+const (
+	// JSONSchemaDraft4URL is the url for JSON Schema Specification Draft 4.
+	JSONSchemaDraft4URL JSONSchemaURL = "http://json-schema.org/draft-04/schema#"
+)
+
+// StringOrArray represents a value that can either be a string or an array of strings.
+// Mainly here for serialization purposes.
+type StringOrArray []string
```

Comment:
I think we have no choice there and must implement it, because we want to decode every JSON schema, but reject our defined openapiv3 subset via validation.